### PR TITLE
Ahelp relay rework, Admin chat relay and the murder of MoMMI OOC bridge

### LINKED ISSUE #19762: Ahelp relay doesn't communicate game mode
If significantly different rules apply to the game mode, it should be clearly communicated in the relay because it can be relevant for ahelps. I don't know what the best way to communicate this is because of the whole game presets vs game rules thing

### LINKED ISSUE #20144: Admin chat is not relayed to discord (and back)
## Description
This should work like OOC chat where game admins can see admin chat from discord and respond to it.

### LINKED ISSUE #23593: Ahelp relay webhook sending discord markdown
## Description
see title
**Reproduction**
Send ahelp with markdown text
Single Ahelp is now annoying

**Screenshots**

image
image

**Additional context**

https://support.discord.com/hc/en-us/articles/210298617-Markdown-Text-101-Chat-Formatting-Bold-Italic-Underline-
Link to discord support guide explaining markdown, includes nice list of things that should be purged from the webhook

### LINKED ISSUE #23716: Make it more obvious when a user is disconnected, maybe change the color of the entire chat's background a bit or something. Alternatively an indicator to sent message to show if the user was online or not when it was sent

### PR DESCRIPTION
**Pull request by Simyon264**
## About the PR
This PR adds a Discord bot to be used in several parts of the codebase. This PR also reworks the ahelp relay to use a forum. See the attached video in the media section for more information. The OOC realay is reworked and the admin chat now also gets a relay. 
## Why / Balance
Resolves #20144. Resolves #19762. Resolves #23593. Resolves #23716. Finishes a TODO from #26954
Also MoMMI is being phased out.

## Technical details
Adds `discord.net` as a NuGet package to `Content.Server`. This is used in `DiscordLink.cs` to host a discord bot which will handle all bot interaction. 

## Media

image
image
https://cdn.discordapp.com/attachments/1182777817105190982/1192456316346712174/Desktop_2024.01.04_-_14.13.43.05.mp4?ex=65a9247c&is=6596af7c&hm=a6c8ff697fc745105511c23f827f36c36a4a0cd64376add81ded79269d0e024a&

(PLEASE SOMEONE TELL ME HOW TO EMBED VIDEO)

image
image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
Removes IMoMMILink, which means existing servers will have to set up the OOC relay again. Should be pretty much a drop in replacement.  This also removes the option to use a webhook for the ahelp relay.
There are 3 CVars needed for the OOC bridge:
`discord.token`: Your Discord bot token.
`discord.guild_id`: The guild ID of your Discord server.
`ooc.relay_channel_id`: The channel ID for the OOC relay.
There are 3 CVars needed for the AHelp relay.
`discord.token`: Your Discord bot token.
`discord.guild_id`: The guild ID of your Discord server.
`admin.ahelp_relay_channel_id`: The FORUM (IT NEEDS TO BE A FORUM CHANNEL!!!) channel ID for the ahelp relay.
There are 3 CVars needed for the admin chat bridge:
`discord.token`: Your Discord bot token.
`discord.guild_id`: The guild ID of your Discord server.
`admin.relay_channel_id`: The channel ID for the admin chat relay.

### CODE REVIEW

**Review by VasilisThePikachu — CHANGES_REQUESTED** [review]
Lil nitpicks

**Review by VasilisThePikachu — COMMENTED** [review]
I don't see the commit where you added this. But the "!status" screenshot example conflicts with redbots (Sussy baka bot on discord) status command. 

I think the fix here is to just make the prefix customizable.

Also ensure it does not do anything if its ran outside a forum channel if you have not already.

**Comment by Simyon264** [comment]
> Also ensure it does not do anything if its ran outside a forum channel if you have not already.

It ignores the command when ran outside of the Ahelp relay forum channel.

File: Content.Server/Discord/DiscordLink.cs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    public DiscordSocketClient? Client;
+    private ISawmill _sawmill = default!;
+
+    private string _guildId = string.Empty;
+
+    public void PostInject()
+    {
+        _sawmill = _logManager.GetSawmill("discord.link");
+
+        Client = new DiscordSocketClient(new DiscordSocketConfig()
```
Comment (PJB3005) [inline Content.Server/Discord/DiscordLink.cs:25]: This should absolutely not be constructed from `PostInject()`. This should be from some initialize call, and only if a token is specified.

File: Content.Server/Discord/DiscordLink.cs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+        try
+        {
+            _configuration.OnValueChanged(CCVars.DiscordGuildId, OnGuildIdChanged, true);
+        }
+        catch (Exception e)
+        {
+            _sawmill.Fatal($"Failed to setup cvar.");
+            return;
+        }
```
Comment (PJB3005) [inline Content.Server/Discord/DiscordLink.cs:39]: Your problem is that you're calling this from `PostInject()`.

File: Content.Server/Discord/DiscordLink.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    private void OnGuildIdChanged(string guildId)
+    {
+        _guildId = guildId;
+    }
+
+    private async Task LoginAsync(string token)
+    {
+        Debug.Assert(Client != null);
+        Debug.Assert(Client.LoginState == LoginState.LoggedOut);
```
Comment (PJB3005) [inline Content.Server/Discord/DiscordLink.cs:78]:
```suggestion
        DebugTools.Assert(Client != null);
        DebugTools.Assert(Client.LoginState == LoginState.LoggedOut);
```

File: Content.Server/Discord/DiscordLink.cs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        return msg.Exception is null ? $"{msg.Source}: {msg.Message}" : $"{msg.Source}: {msg.Exception}\n{msg.Message}";
+    }
+
+    private Task Log(LogMessage msg)
+    {
+        switch (msg.Severity)
+        {
+            case LogSeverity.Critical:
+                _sawmill.Fatal(FormatLog(msg));
```
Comment (PJB3005) [inline Content.Server/Discord/DiscordLink.cs:97]: Please provide a separate sawmill for relaying messages created by Discord.Net. Something like `discord.link.log`.

File: Content.Server/Discord/DiscordLink.cs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+            return;
+        }
+
+        if (_guildId == string.Empty)
+        {
+            _sawmill.Error("Tried to send a Discord message but the guild ID is not set! Blow up now!");
+            return;
+        }
+
+        Client.GetGuild(ulong.Parse(_guildId))
```
Comment (PJB3005) [inline Content.Server/Discord/DiscordLink.cs:127]: This `ulong` should be parsed when the CVar is read from the config.

File: Content.Server/Discord/DiscordLink.cs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    private string FormatLog(LogMessage msg)
+    {
+        return msg.Exception is null ? $"{msg.Source}: {msg.Message}" : $"{msg.Source}: {msg.Exception}\n{msg.Message}";
+    }
+
+    private Task Log(LogMessage msg)
+    {
+        switch (msg.Severity)
```
Comment (PJB3005) [inline Content.Server/Discord/DiscordLink.cs:94]: Use `_sawmill.Log` as it allows you to pass the log level as a parameter, then use a switch expression to translate the log levels.

File: Content.Server/Discord/DiscordLink.cs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        }
+        catch (Exception e)
+        {
+            _sawmill.Fatal($"Failed to setup cvar.");
+            return;
+        }
+
+        if (_configuration.GetCVar(CCVars.DiscordToken) is not { } token)
+        {
+            _sawmill.Warning("No Discord token specified, not connecting.");
```
Comment (PJB3005) [inline Content.Server/Discord/DiscordLink.cs:43]: This should not be a warning case, it is a totally valid use for game servers to not have the Discord integration configured.

File: Content.Server/Discord/DiscordLink.cs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    private async Task LoginAsync(string token)
+    {
+        Debug.Assert(Client != null);
+        Debug.Assert(Client.LoginState == LoginState.LoggedOut);
+
+        await Client.LoginAsync(TokenType.Bot, token);
+        await Client.StartAsync();
+
+        _sawmill.Info("Connected to Discord.");
+        await Task.Delay(-1); // Following discord.net guide, idk what im doiung help
```
Comment (PJB3005) [inline Content.Server/Discord/DiscordLink.cs:84]: You don't need to do this, the game server will keep the process running.

File: Content.Shared/CCVar/CCVars.cs [hunk 9]
```diff
@@ -397,7 +387,23 @@ public sealed class CCVars : CVars
         public static readonly CVarDef<string> DiscordRoundEndRoleWebhook =
             CVarDef.Create("discord.round_end_role", string.Empty, CVar.SERVERONLY);
 
+        /// <summary>
+        ///     Discord bot token for link stuff.
+        /// </summary>
```
Comment (PJB3005) [inline Content.Shared/CCVar/CCVars.cs:392]: This entire file needs better comments indicating that these various CVars you added are linked and need each other to be configured.

File: Content.Shared/CCVar/CCVars.cs [hunk 10]
```diff
@@ -397,7 +387,23 @@ public sealed class CCVars : CVars
...
+            CVarDef.Create("discord.token", string.Empty, CVar.SERVERONLY | CVar.CONFIDENTIAL);
 
+        /// <summary>
+        ///     The Discord guild ID to use for link stuff.
+        /// </summary>
+        public static readonly CVarDef<string> DiscordGuildId =
+            CVarDef.Create("discord.guild_id", string.Empty, CVar.SERVERONLY);
+
+        /// <summary>
+        ///     Prefix for the Discord bot.
```
Comment (PJB3005) [inline Content.Shared/CCVar/CCVars.cs:403]: Indicate that this is a prefix for all the _commands_.

**Comment by IProduceWidgets** [comment]
Ideally players should not see that the admin is not in game.

**Comment by Peptide90** [comment]
> Ideally players should not see that the admin is not in game.

People admin checking is as old as time and an ahelp will do it. They'll find a way, but yeah if it's easy enough to hide sure.

**Comment by Simyon264** [comment]
Whoops forgot to show it in the screenshot! Admins who reply using the ahelp relay are marked yellow and have a little symbol in front of their name.

**Comment by metalgearsloth** [comment]
@Simyon264 is this still being worked on.

**Comment by Simyon264** [comment]
Yes. I have been taking a short break for about a week now.

**Comment by Simyon264** [comment]
> Ideally players should not see that the admin is not in game.

I added a CVar to hide that the message was sent from discord, but the Relay uses your discord username, which means it's easy to check if it was sent from discord by just matching the ingame username from the discord username (Also Discord usernames are all lowercase).

**Comment by Simyon264** [comment]
Test fails seems unrelated.

File: Content.Shared/CCVar/CCVars.cs [hunk 11]
```diff
@@ -773,6 +770,18 @@ public sealed class CCVars : CVars
...
+        ///     The channel that admin messages are sent to.
+        /// </summary>
+        public static readonly CVarDef<string> AdminRelayChannelId =
+            CVarDef.Create("admin.relay_channel_id", string.Empty, CVar.SERVERONLY);
+
+        /// <summary>
+        ///     The **FORUM** channel that admin messages are sent to. If it's not a forum channel, everything will explode.
+        /// </summary>
+        public static readonly CVarDef<string> AdminAhelpRelayChannelId =
+            CVarDef.Create("admin.ahelp_relay_channel_id", string.Empty, CVar.SERVERONLY);
```
Comment (ShadowCommander) [inline Content.Shared/CCVar/CCVars.cs:783]: Mention Discord so that it's clearer what these are for.

File: Content.Shared/CCVar/CCVars.cs [hunk 12]
```diff
@@ -1219,6 +1235,12 @@ public sealed class CCVars : CVars
         public static readonly CVarDef<bool> OocEnableDuringRound =
             CVarDef.Create("ooc.enable_during_round", false, CVar.NOTIFY | CVar.REPLICATED | CVar.SERVER);
 
+        /// <summary>
+        /// The channel that OOC messages are sent to.
```
Comment (ShadowCommander) [inline Content.Shared/CCVar/CCVars.cs:1239]: Add Discord mention

File: Content.Shared/CCVar/CCVars.cs [hunk 13]
```diff
@@ -860,12 +869,19 @@ public sealed class CCVars : CVars
             CVarDef.Create("admin.afk_time", 600f, CVar.SERVERONLY);
 
         /// <summary>
+        ///     If this is true, the ahelp relay shows that the response was from discord. If this is false, all messages an admin sends will be shown as if the admin was ingame.
+        /// </summary>
+        /// <returns></returns>
+        public static readonly CVarDef<bool> AdminAhelpRelayShowDiscord =
+            CVarDef.Create("admin.ahelp_relay_show_discord", true, CVar.SERVERONLY);
```
Comment (ShadowCommander) [inline Content.Shared/CCVar/CCVars.cs:876]:
```suggestion
        ///     If this is true, the ahelp relay shows that the response was from Discord. If this is false, all messages an admin sends will be shown as if the admin was ingame.
        /// </summary>
        public static readonly CVarDef<bool> AdminAhelpRelayShowDiscord =
            CVarDef.Create("admin.ahelp_relay_show_discord", true, CVar.SERVERONLY);
```

**Review by ShadowCommander — CHANGES_REQUESTED** [review]

**Comment by Simyon264** [comment]
Reviews done.

**Review by ShadowCommander — APPROVED** [review]
Looks good, though I haven't done a super in depth review.

**Comment by Simyon264** [comment]
Don't worry, PJB will do a nuclear bomb shell of a review and I'm ready for it.

File: Content.Server/Administration/Systems/BwoinkSystem.cs [hunk 14]
```diff
@@ -1,17 +1,19 @@
...
 using Content.Server.Discord;
 using Content.Server.GameTicking;
+using Content.Server.Maps;
+using Content.Shared.Access.Systems;
 using Content.Shared.Administration;
 using Content.Shared.CCVar;
 using Content.Shared.Mind;
+using Discord;
+using Discord.Rest;
+using Discord.WebSocket;
```
Comment (PJB3005) [inline Content.Server/Administration/Systems/BwoinkSystem.cs:16]: There should be no references to Discord.Net in this file.

In fact, ideally there wouldn't even be references to *Discord* in this file. There should be a separate manager in charge of bridging bwoinks to Discord. You'd have:
* `BwoinkSystem` (this shouldn't be a system but that's not your fault)
* `DiscordLink`
* `DiscordBwoinkLink`
* `DiscordChatLink`

File: Content.Server/Chat/Managers/ChatManager.cs [hunk 15]
```diff
@@ -63,10 +69,54 @@ public void Initialize()
...
             _configurationManager.OnValueChanged(CCVars.AdminOocEnabled, OnAdminOocEnabledChanged, true);
+            _configurationManager.OnValueChanged(CCVars.OocRelayChannelId, OnOocChannelChanged, true);
+            _configurationManager.OnValueChanged(CCVars.AdminRelayChannelId, OnAdminChannelChanged, true);
+
+            _discordLink.OnMessageReceived += MessageReceived;
 
             _playerManager.PlayerStatusChanged += PlayerStatusChanged;
         }
 
+        private void MessageReceived(SocketMessage arg)
```
Comment (PJB3005) [inline Content.Server/Chat/Managers/ChatManager.cs:80]: This functionality shouldn't be in chat manager.

File: Content.Server/Administration/Systems/BwoinkSystem.cs [hunk 16]
```diff
@@ -499,25 +670,30 @@ private IList<INetChannel> GetTargetAdmins()
...
-                stringbuilder.Append($" **{roundTime}**");
             if (!playedSound)
                 stringbuilder.Append(" **(S)**");
+
             stringbuilder.Append($" **{username}:** ");
             stringbuilder.Append(message);
             return stringbuilder.ToString();
         }
+
+        private static string EscapeMarkdown(string text)
```
Comment (PJB3005) [inline Content.Server/Administration/Systems/BwoinkSystem.cs:693]: Does Discord.Net not have functionality for this?

File: Content.Server/Administration/Systems/BwoinkSystem.cs [hunk 17]
```diff
@@ -147,214 +324,215 @@ private void OnServerNameChanged(string obj)
...
         }
 
-        private WebhookPayload GeneratePayload(string messages, string username, string? characterName = null)
+        private EmbedBuilder GenerateEmbed(ICommonSession session)
         {
-            // Add character name
-            if (characterName != null)
-                username += $" ({characterName})";
+            var clr = GetNonAfkAdmins().Count > 0 ? Color.Green : Color.Red;
+            var job = "No entity attached";
```
Comment (PJB3005) [inline Content.Server/Administration/Systems/BwoinkSystem.cs:444]: These need to be localized strings

File: Content.Server/Administration/Systems/BwoinkSystem.cs [hunk 18]
```diff
@@ -147,214 +324,215 @@ private void OnServerNameChanged(string obj)
...
+                        null,
+                        embed.Build()
+                    );
+
+            // If the relay is not null, give a list of available commands for the admins to use.
+            if (relay is not null)
             {
-                var request = await _httpClient.PatchAsync($"{_webhookUrl}/messages/{existingEmbed.id}",
-                    new StringContent(JsonSerializer.Serialize(payload), Encoding.UTF8, "application/json"));
+                await relay.SendMessageAsync($"Use `{_discord.BotPrefix}status` to regenerate the status embed as seen above. `-` in the beginning of a message will be cause the message to not be relayed to the player.");
```
Comment (PJB3005) [inline Content.Server/Administration/Systems/BwoinkSystem.cs:428]: Localized string

**Review by PJB3005 — CHANGES_REQUESTED** [review]
Try again

**Comment by metalgearsloth** [comment]
derelict.

**Comment by Simyon264** [comment]
I will come back to this once #28639 is merged as it touches the BwoinkSystem as well.

**Comment by Simyon264** [comment]
I am quite short on time in the following days (mainly because of my upcoming birthday). I will reopen this once I find the time.

**Comment by AwareFoxy** [comment]
> I am quite short on time in the following days (mainly because of my upcoming birthday). I will reopen this once I find the time.

hi, can I take care of this if you don't want to go back to it

**Comment by Simyon264** [comment]
> > I am quite short on time in the following days (mainly because of my upcoming birthday). I will reopen this once I find the time.
> 
> hi, can I take care of this if you don't want to go back to it

I now have the time and will soon atomize this PR into two distinct issues.
1: The ahelp relay
2: The ooc relay.

I will first do the ooc relay as that is much simpler and integrates a bot into Content.

**Comment by JIPDawg** [comment]
Guten tag, 

give pl0x.

Auf Wiedersehn.

**Pull request closed without merging**